sdk/metric/controller/basic/controller.go

```diff
@@ -254,34 +293,41 @@ func (c *Controller) checkpoint(ctx context.Context, cond func() bool) error {
 	return err
 }
 
-// export calls the exporter with a read lock on the CheckpointSet,
+// export calls the exporter with a read lock on the Reader,
 // applying the configured export timeout.
 func (c *Controller) export(ctx context.Context) error {
-	ckpt := c.checkpointer.CheckpointSet()
-	ckpt.RLock()
-	defer ckpt.RUnlock()
-
 	if c.pushTimeout > 0 {
 		var cancel context.CancelFunc
 		ctx, cancel = context.WithTimeout(ctx, c.pushTimeout)
 		defer cancel()
 	}
 
-	return c.exporter.Export(ctx, c.resource, ckpt)
+	return c.exporter.Export(ctx, c.resource, c.Reader())
 }
 
-// ForEach gives the caller read-locked access to the current
-// export.CheckpointSet.
-func (c *Controller) ForEach(ks export.ExportKindSelector, f func(export.Record) error) error {
-	ckpt := c.checkpointer.CheckpointSet()
-	ckpt.RLock()
-	defer ckpt.RUnlock()
+type libraryReader struct {
+	*Controller
+}
 
-	return ckpt.ForEach(ks, f)
+var _ export.InstrumentationLibraryReader = libraryReader{}
+
+func (l libraryReader) ForEach(readerFunc func(l instrumentation.Library, r export.Reader) error) error {
+	for _, impl := range l.Controller.provider.List() {
+		pair := impl.(*accumulatorCheckpointer)
```

Comment:
Similar here, should this panic if the type assertion fails? Do we want to skip the impl and log an error instead? Is there a way we could design this so the type assertion isn't necessary?

Reply:
Fixed with "impossible type assertion failure" error.

> Is there a way we could design this so the type assertion isn't necessary?

Yes, but it would lead to duplication of the registry code, which is used by the global and by the real metric controller. If we had generics, the registry would use them and this would not require a type assertion.

If we copy the registry code into at least two places, there wouldn't be a type assertion, but that doesn't fee like a win. The registry code is responsible for ensuring that duplicate instruments are registered with consistent types, so both the real and global implementations are supposed to do this.